Updated contributing.md and the style guide (#670)
mosaicml/composer: CONTRIBUTING.md

1. Suppose a variable could be one of multiple types, like the following:

    ```python
    def foo(x: Union[int, None]):
        """
        Args:
            x (int or None): Foo parameter
        """
        return x + 5  # type error -- None + 5 is not allowed!
    ```

    PyRight will complain since `None + 5` is not a valid operation.
    Instead, add a check to ensure that `x is not None`:

    ```python
    def foo(x: Union[int, None]):
        if x is None:
            raise TypeError("x must be an integer, not None!")
        return x + 5  # valid
    ```

    Assert statements also work. However, assert statements should NOT be used for data validation
    (see the assert statement section below).
    ```python
    def foo(x: Union[int, None]):
        assert x is not None, "x should never be None"
        return x + 5  # valid
    ```

1. For variables where it is impossible for PyRight to infer the correct type, use
[cast](https://docs.python.org/3/library/typing.html#typing.cast).
1. As a last resort, add a `# type: ignore` comment to the line where PyRight emits an error.
Immediately following this statement, paste in the error emitted by PyRight,

> so other contributors will know why this error was silenced.

Author: I prefer to include the full error message. Does Pyright support the bracket notation?

## 4. Public APIs
A public API, generally speaking, can be invoked by a user without a leading underscore in any portion of the path.
The following are examples of public APIs:

* Standalone functions in public modules (e.g. `composer.utils.dist.get_world_size`)
* Classes in public modules (e.g. `composer.trainer.trainer.Trainer`)
* Public methods in public classes (e.g. `composer.trainer.trainer.Trainer.fit`)
* Public modules (e.g. `composer.trainer.trainer`)

The following rules apply to public APIs:
1. All public APIs must have a docstring (see the Documentation section below)
1. All parameters must have type annotations.
1. Parameters should be flattened, rather than nested. While this violates the software engineering practice of
    being DRY (don't repeat yourself), it simplifies the user API by making all parameters visible.
1. To minimize user imports, parameters should support native PyTorch types or Python primitives whenever possible.

> 

Author: Agree with mapping over dict (though note that mapping is immutable, so a mutablemapping might need to be used sometimes). Re sequences, that will depend on #677.

from composer.core.callback import Callback

log = logging.getLogger(__name__)

__all__ = ["MemoryMonitor"]  # export only the MemoryMonitor, not other imports like `Logger`, `State`, or `Callback`

class MemoryMonitor(Callback):
    ...
```

### 6.3 `__init__.py`

All public classes and functions should be added to the module's `__init__.py`.

```python
from composer.path.to.module.file import MyClass as MyClass
from composer.path.to.module.file import my_func as my_func
```

If a file only contains public functions, then the following is also acceptable:

```python
from composer.path.to.module import my_file as my_file
```

## 7. Documentation

Composer uses [Google Style Docstrings](https://sphinxcontrib-napoleon.readthedocs.io/en/latest/example_google.html).

The following guidelines apply to documentation.
1.  Each function that needs a docstring must have its input arguments and return statement (if not None) annotated.
1.  Each argument annotation should include the type. If the argument has a default value, the type annotation should
    specify "optional", and the docstring should say the default value. There should only be one line for the "Returns:"

>     section, even if a tuple is returned. For example:

Author: Ah good catch; didn't realize google style supported this. Agree here; let's do it this way then.

    ```python

    def foo(bar: int):

>         """Foo.

Reviewer: Can these one-liners be examples of using imperative rather than descriptive style, at least if we care about that? E.g.:
`"""Add 5 to the input"""`
as opposed to:
`"""Adds 5 to the input"""`

        Args:
            bar (int): Required bar.
        """
        ...

    def foo2(bar: int = 42):
        """Foo2.

        Args:
            bar (int, optional): Optional bar. (default: ``42``)
        """
        ...
    
    def foo3(bar: Optional[int] = None):
        """Foo3.

        Args:

>             bar (int, optional): Optional bar. (default: ``None``)

Reviewer 2: I may have been mistaken when I effected that in #662 as our Trainer does (default: `42`).

    `.. testcode::` . See the [Sphinx Doctest Extension](https://www.sphinx-doc.org/en/master/usage/extensions/doctest.html)
    for all of the available directives. Do not use `.. code-block::` for Python examples, as they are untested.

    Any test fixtures for doctests should go in [docs/source/doctest_fixtures.py](docs/source//doctest_fixtures.py)
    or in a `.. testsetup::` block.

    For example:
    ```python
    def my_function(x: Optional[torch.Tensor]) -> torch.Tensor:
        """blah function

        Args:
            input (torch.Tensor, optional): Your guess.

        Returns:
            torch.Tensor: How good your input is.

        Raises:
            ValueError: If your input is negative.

        Example:
            .. testsetup::

                # optional setup section, not shown in docs
                import torch
                x = torch.randn(42)
                

            .. testcode::

                # shown in docs; runs after testsetup
                my_function(x)
        """
        ...
    ```

    To check doctests, run:
    
    ```bash
    cd docs && make clean && make html && make doctest

>     ```

Reviewer: already mentioned in section on running the tests. also we should probably have `make doctest` in the root directory run the doctests in one command, but that's a separate issue

## 8. Hparams and Config Classes

Each class that could be initialized by the hparams will need its configuration variables defined in

> [yahp](https://github.com/mosaicml/yahp) dataclass.

Reviewer: Not sure what this sentence means. What does it mean to be "initialized by the hparams"? How do I know if a class I write could be initialized in this way? also missing "a" before "yahp".

Also, will future code have an associated yahp dataclass? If not, can maybe just rm this section

Guidelines:
* The fields in the dataclass should roughly follow the init signature of the class being constructed
* `initialize_object(self)` should take any parameters needed to construct the class. It acceptable to

> take complex types or other hparam dataclasses, as required, to initialize the object.

Reviewer: "It acceptable" -> "It is acceptable"

Issues tagged with [good first issue](https://github.com/mosaicml/composer/issues?q=is%3Aissue+is%3Aopen+label%3A%22good+first+issue%22) are great options to start contributing.

If you have questions, join us on [Slack](https://join.slack.com/t/mosaicml-community/shared_invite/zt-w0tiddn9-WGTlRpfjcO9J5jyrMub1dg) -- we'll be happy to help you!

We welcome contributions for bug fixes, new efficient methods you'd like to contribute to the community, or new models and datasets!

## New Algorithms

> Have a new algorithm you'd like to contribute to the library as part of your research? Reach out to us on Slack or file a new issue with the proposed method, and we'll work with you to get started!

Reviewer: What if people just built it for themselves and want to open a PR directly? "we'll work with you to get started" sounds 1) it's hard and 2) there's a hurdle to jump through, sort of like when a pricing page says "get in touch" instead of a price.

We can always reject PRs that aren't up to our standards, so I'd vote for saying new method PRs are welcome, but we suggest opening an issue first to discuss how and whether it makes sense to do it.

```bash
pip install -e .[dev]
```

We use a few formatters for code style, and you run the following to autocorrect your files:
```bash
make style
```

That will run the [yapf](https://github.com/google/yapf) formatter for general formatting,
[isort](https://github.com/PyCQA/isort) to sort imports, and
[docformatter](https://github.com/myint/docformatter) to format docstrings.

1. When you are ready, submit a pull request into the composer repository! If merged, we'll reach out to send you some free swag :)

## Running Tests

To test your changes locally, run:

1. `make test`  # run CPU tests
1. `make test-gpu`  # run GPU tests
1. `cd docs && make doctest`  # run doctests

Some of our checks test distributed training as well. To test these, run:

* `make test-dist WORLD_SIZE=2`  # run 2-cpu distributed tests
* `make test-dist-gpu WORLD_SIZE=2`  # run 2-gpu distributed tests

These tests run with the `composer` launcher. We also support `WORLD_SIZE=1`, which would run the tests with the `composer` launcher on a single device.

See the [Makefile](https://github.com/mosaicml/composer/blob/dev/Makefile) for more information.

## Code Style & Typing

Follow Google's

> [Python Style Guide](https://google.github.io/styleguide/pyguide.html) for how to format and structure code. Many of these are already taken care of by the `make style` command above.

Reviewer: "these" -> "these guidelines", since "these" has unclear antecedent

Composer aims to annotate all functions with type annotations (introduced in

> [PEP 526](https://www.python.org/dev/peps/pep-0526/). Not to worry if you are not a python typing expert, put in the pull request, and we'll help you with getting the code into shape.

Reviewer: ```suggestion
[PEP 526](https://www.python.org/dev/peps/pep-0526/)). Don't worry if you are not a Python typing expert; put in the pull request and we'll help you get the code into shape.
```

Found this a little hard to parse because of the parens wrapping end of one sentence and then next sentence.
